txpipe/extensions/twopoint_scia.py

```diff
@@ -0,0 +1,493 @@
+        ng.process(cat_j, cat_i)
+
+        if rancat_j:
+            rg = treecorr.NGCorrelation(self.config)
+            rg.process(rancat_j, cat_i)
+        else:
+            rg = None
+
+        ng.calculateXi(rg=rg)
+
+        return ng
+
+    def call_treecorr(self, data, meta, i, j, k):
+        import sacc 
+
+        if k==SHEAR_SHEAR:
+            xx = self.calculate_shear_shear(data, meta, i, j)
+            xtype = "combined"
+        elif k==SHEAR_POS:
+            xx = self.calculate_shear_pos(data, meta, i, j)
+            xtype = sacc.standard_types.galaxy_shearDensity_xi_t
+        elif k==SHEAR_POS_SELECT:
+            xx = self.calculate_shear_pos_select(data, meta, i, j)
+            xtype = sacc.build_data_type_name('galaxy',['shear','Density'],'xi',subtype ='ts')
+        elif k==POS_POS:
+            xx = self.calculate_pos_pos(data, meta, i, j)
+            xtype = sacc.standard_types.galaxy_density_xi
+        else:
+            raise ValueError(f"Unknown correlation function {k}")
+
+        result = Measurement(xtype, xx, i, j)
+
+        sys.stdout.flush()
+        return result
+
+    def write_output(self, data, meta, results):
+        import sacc
+        import treecorr
+        XI = "combined"
+        XIP = sacc.standard_types.galaxy_shear_xi_plus
+        XIM = sacc.standard_types.galaxy_shear_xi_minus
+        GAMMAT = sacc.standard_types.galaxy_shearDensity_xi_t
+        GAMMATS = sacc.build_data_type_name('galaxy',['shear','Density'],'xi',subtype ='ts')
+        WTHETA = sacc.standard_types.galaxy_density_xi
+        GAMMAX = sacc.standard_types.galaxy_shearDensity_xi_x
+        GAMMAXS = sacc.build_data_type_name('galaxy',['shear','Density'],'xi',subtype ='xs')
+
+        S = sacc.Sacc()
+        if self.config['do_shear_pos'] == True:
+            S2 = sacc.Sacc()
+
+        # We include the n(z) data in the output.
+        # So here we load it in and add it to the data
+        f = self.open_input('shear_photoz_stack')
+
+        # Load the tracer data N(z) from an input file and
+        # copy it to the output, for convenience
+        for i in data['source_list']:
+            z = f['n_of_z/source/z'][:]
+            Nz = f[f'n_of_z/source/bin_{i}'][:]
+            S.add_tracer('NZ', f'source_{i}', z, Nz)
+            if self.config['do_shear_pos'] == True:
+                S2.add_tracer('NZ', f'source_{i}', z, Nz)
+
+
+        f.close()
+
+        # Now build up the collection of data points, adding them all to
+        # the sacc data one by one.
+        comb = []
+        for d in results:
+            # First the tracers and generic tags
+            tracer1 = f'source_{d.i}' #if d.corr_type in [XI, GAMMAT,GAMMATS, ] else f'lens_{d.i}'
+            tracer2 = f'source_{d.j}' #if d.corr_type in [XI, GAMMAT, GAMMATS] else f'lens_{d.j}'
+
+            # We build up the comb list to get the covariance of it later
+            # in the same order as our data points
+            comb.append(d.object)
```

Comment:
would be nice to a few brief comments like this as needed in the self calibration specific pieces of code with the selection function

Reply:
Tried adding a few comments.